proposals/open-telemetry/microprofile-opentelemetry-tracing.asciidoc

```diff
@@ -0,0 +1,190 @@
+:authors: Roberto Cortez
+:email: radcortez@yahoo.com
+:version-label!:
+:sectanchors:
+:doctype: book
+:license: Eclipse Foundation Specification License v1.0
+:source-highlighter: coderay
+:toc: left
+:toclevels: 4
+:sectnumlevels: 4
+ifdef::backend-pdf[]
+:pagenums:
+endif::[]
+
+= Eclipse MicroProfile OpenTelemetry Tracing
+
+== Introduction
+
+TODO
+
+== Rationale
+
+TODO
+
+== Architecture
+
+This specification defines an easy way to allow an application running in a cloud native container to take advantage of
+distributed tracing by using the https://opentelemetry.io[OpenTelemetry project] (a merger of the OpenTracing and
+OpenCensus projects).
+
+This document and implementations *MUST* comply with the following OpenTelemetry specifications:
+
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/overview.md[OpenTelemetry Overview]
+(except for Metrics)
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md[Tracing API]
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/baggage/api.md[Baggage API]
+- https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/context[Context API]
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md[Resource SDK]
+
+=== Automatic Instrumentation
+
+JAX-RS applications, JAX-RS Clients, and MicroProfile REST Clients are automatically enlisted to participate in
+distributed tracing without code modification.
+
+=== Manual Instrumentation
+
+Explicit manual instrumentation can be added into a MicroProfile application in the following ways:
+
+==== @WithSpan
+
+Annotating a method in any Jakarta CDI aware beans with the `io.opentelemetry.extension.annotations.WithSpan`
+annotation. This will create a new Span and establish any required relationships with the current Trace context.
+
+Method parameters can be annotated with the `io.opentelemetry.extension.annotations.SpanAttribute` annotation to
+indicate which method parameters should be part of the Trace.
+
+Example:
+[source,java]
+----
+@ApplicationScoped
+class SpanBean {
+    @WithSpan
+    void span() {
+
+    }
+
+    @WithSpan("name")
+    void spanName() {
+
+    }
+
+    @WithSpan(kind = SERVER)
+    void spanKind() {
+
+    }
+
+    @WithSpan
+    void spanArgs(@SpanAttribute(value = "arg") String arg) {
+
+    }
+}
+----
+
+==== Obtain a SpanBuilder
+
+By obtaining a `SpanBuilder` from the current `Tracer` and calling
+`io.opentelemetry.api.trace.Tracer.spanBuilder(String)`. In this case, it is the developer's responsibility to ensure
+that the `Span` is properly created, closed, and propagated.
+
+Example:
+[source,java]
+----
+@RequestScoped
+@Path("/")
+public class SpanResource {
+    @Inject
+    Tracer tracer;
+
+    @GET
+    @Path("/span/new")
+    public Response spanNew() {
```

Comment:
This example is a bit confusing.  The automatic JAX-RS tracing will create a span for `/span/new`.  Any new span created under a current span will be a child of the current span by default.  We will end up having 2 spans here.  Is this the intent?

Reply:
Yes. This is just an example if the user wants to create a new `Span` (and not reuse the current one). I can clarify the text.